File: src/main/java/com/embabel/prepper/agent/ContactService.java
package com.embabel.prepper.agent;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Optional;

@Service
public class ContactService {

    private final Logger logger = LoggerFactory.getLogger(ContactService.class);
    private final ResearchedParticipantRepository repository;

    public ContactService(ResearchedParticipantRepository repository) {
        this.repository = repository;
    }

    @Transactional(readOnly = true)
    public Optional<Domain.Contact> resolveContact(String identification) {
        // TODO real implementation might use others means of identification,
        // including full text search, phone number, etc,
        // but we rely on email
        var found = repository.findByEmail(identification);
        logger.info("Resolved contact for {}: {}", identification, found);
        return found;
    }

    @Transactional
    public Domain.Contact createContact(Domain.NewContact newContact) {
        var saved = repository.save(new Domain.Contact(newContact));
        logger.info("Created new contact: {}", saved);
        return saved;
    }
}

File: src/main/kotlin/com/embabel/prepper/agent/ContactService.kt
package com.embabel.prepper.agent

import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class ContactService(private val repository: ResearchedParticipantRepository) {

    private val logger = LoggerFactory.getLogger(ContactService::class.java)

    @Transactional(readOnly = true)
    fun resolveContact(identification: String) = repository.findByEmail(identification).also {
        logger.info("Resolved contact for {}: {}", identification, it)
    }

    @Transactional(readOnly = true)
    fun findAll(): List<Domain.Contact> = repository.findAll()

    @Transactional
    fun createContact(newContact: Domain.NewContact): Domain.Contact {
        val saved = repository.save(Domain.Contact(newContact))
        logger.info("Created new contact: {}", saved)
        return saved
    }
}
